File: fix-a6a7.py
#!/usr/bin/env python3
"""
Fix A6 (answer distribution >5) and A7 (3 consecutive same answers) violations.

Strategy:
1. For A6/A7 on non-marker questions: swap choices within the question
2. For A7 on marker-only runs: swap question POSITIONS with a non-run question
   that has a different ans value
3. For A6 when only marker questions have the over-represented ans:
   swap a marker question's position with a swappable question of different ans,
   then adjust the swappable question's ans if needed
"""

import json
import sys
import os
import subprocess
import re
import copy

def is_marker(q):
    """Check if question is marker type (choices are ①②③④)."""
    return q.get('ch', []) == ['①', '②', '③', '④']

def is_swappable(q):
    """Check if a question's choices can be swapped (reordered)."""
    ch = q.get('ch', [])
    if ch == ['①', '②', '③', '④']:
        return False
    if ch == ['T', 'F']:
        return False
    if len(ch) == 0:
        return False
    if len(ch) != 4:
        return False
    ans = q.get('ans')
    if not isinstance(ans, int):
        return False
    return True

def get_ans_distribution(questions):
    dist = {1: 0, 2: 0, 3: 0, 4: 0}
    for q in questions:
        ans = q.get('ans')
        if isinstance(ans, int) and ans in dist:
            dist[ans] += 1
    return dist

def find_a7_runs(questions):
    """Find first A7 violation (3 consecutive same int answers)."""
    ans_list = [q.get('ans') for q in questions]
    for i in range(len(ans_list) - 2):
        a = ans_list[i]
        if isinstance(a, int) and a == ans_list[i+1] == ans_list[i+2]:
            return (i, i+1, i+2, a)
    return None

def would_create_a7(questions, idx, new_ans):
    """Check if changing questions[idx].ans to new_ans creates A7."""
    if not isinstance(new_ans, int):
        return False
    n = len(questions)
    if idx >= 2:
        if questions[idx-2].get('ans') == questions[idx-1].get('ans') == new_ans:
            return True
    if idx >= 1 and idx + 1 < n:
        if questions[idx-1].get('ans') == new_ans and questions[idx+1].get('ans') == new_ans:
            return True
    if idx + 2 < n:
        if questions[idx+1].get('ans') == new_ans and questions[idx+2].get('ans') == new_ans:
            return True
    return False

def would_position_swap_create_a7(questions, idx_a, idx_b):
    """Check if swapping positions of questions at idx_a and idx_b creates new A7."""
    # Simulate the swap
    qs = list(questions)
    qs[idx_a], qs[idx_b] = qs[idx_b], qs[idx_a]
    # Check all positions around both swap points
    for check_idx in range(len(qs)):
        ans = qs[check_idx].get('ans')
        if not isinstance(ans, int):
            continue
        if check_idx >= 2:
            if qs[check_idx-2].get('ans') == qs[check_idx-1].get('ans') == ans:
                return True
    return False

def swap_choices(q, old_ans_idx, new_ans_idx):
    """Swap two choices and update ans and det.analysis."""
    q = copy.deepcopy(q)
    ch = q['ch']
    ch[old_ans_idx], ch[new_ans_idx] = ch[new_ans_idx], ch[old_ans_idx]
    q['ans'] = new_ans_idx + 1

    det = q.get('det', {})
    if 'analysis' in det:
        analysis = det['analysis']
        markers = ['①', '②', '③', '④']
        old_m = markers[old_ans_idx]
        new_m = markers[new_ans_idx]

        lines = analysis.split('\n')
        old_li = new_li = None
        for li, line in enumerate(lines):
            s = line.strip()
            if s.startswith(old_m):
                old_li = li
            elif s.startswith(new_m):
                new_li = li

        if old_li is not None and new_li is not None:
            old_line = lines[old_li]
            new_line = lines[new_li]
            old_content = old_line.split(old_m, 1)[1] if old_m in old_line else ''
            new_content = new_line.split(new_m, 1)[1] if new_m in new_line else ''
            arrow = ' ←정답'
            old_content = old_content.replace(arrow, '')
            new_content = new_content.replace(arrow, '')
            lines[old_li] = old_m + new_content
            lines[new_li] = new_m + old_content + arrow
            det['analysis'] = '\n'.join(lines)

    q['det'] = det
    return q

def pick_best_target(current_ans, dist, forbidden):
    candidates = [(dist.get(n, 0), n) for n in [1,2,3,4] if n != current_ans and n not in forbidden]
    candidates.sort()
    return candidates[0][1] if candidates else None

def fix_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        data = json.load(f)

    questions = data['questions']
    modified = False

    for iteration in range(50):
        dist = get_ans_distribution(questions)
        a7_run = find_a7_runs(questions)
        a6_over = {k: v for k, v in dist.items() if v > 5}

        if not a6_over and not a7_run:
            break

        fixed_one = False

        # --- Fix A7 first ---
        if a7_run:
            ri, rj, rk, run_val = a7_run

            # Strategy 1: Find swappable question IN the run and change its ans
            for qi in [rj, ri, rk]:  # prefer middle
                if is_swappable(questions[qi]):
                    forbidden = {run_val}
                    for c in [1,2,3,4]:
                        if c != run_val and would_create_a7(questions, qi, c):
                            forbidden.add(c)
                    target = pick_best_target(run_val, dist, forbidden)
                    if target:
                        questions[qi] = swap_choices(questions[qi], run_val - 1, target - 1)
                        modified = True
                        fixed_one = True
                        break

            # Strategy 2: Swap POSITIONS - put a different-ans question into the run
            if not fixed_one:
                # Find the middle of the run
                swap_in_idx = rj  # middle position to break

                # Find a question OUTSIDE the run with different ans
                # Prefer questions with different ans that won't create new A7
                for qi in range(len(questions)):
                    if qi in [ri, rj, rk]:
                        continue
                    q_ans = questions[qi].get('ans')
                    if not isinstance(q_ans, int) or q_ans == run_val:
                        continue

                    # Check if swapping positions would create new A7
                    if not would_position_swap_create_a7(questions, swap_in_idx, qi):
                        # Do the position swap
                        questions[swap_in_idx], questions[qi] = questions[qi], questions[swap_in_idx]
                        modified = True
                        fixed_one = True
                        break

            # Strategy 3: Swap positions with ANY different-ans question
            if not fixed_one:
                for qi in range(len(questions)):
                    if qi in [ri, rj, rk]:
                        continue
                    q_ans = questions[qi].get('ans')
                    if not isinstance(q_ans, int) or q_ans == run_val:
                        continue
                    # Force swap even if it might create A7 (we'll fix it next iteration)
                    questions[rj], questions[qi] = questions[qi], questions[rj]
                    modified = True
                    fixed_one = True
                    break

            if fixed_one:
                continue

        # --- Fix A6 ---
        if a6_over:
            over_ans = max(a6_over, key=a6_over.get)

            # Strategy 1: Find a swappable question with over_ans and change it
            for qi in range(len(questions)):
                if questions[qi].get('ans') == over_ans and is_swappable(questions[qi]):
                    forbidden = set()
                    for c in [1,2,3,4]:
                        if c != over_ans and would_create_a7(questions, qi, c):
                            forbidden.add(c)
                    target = pick_best_target(over_ans, dist, forbidden)
                    if target:
                        questions[qi] = swap_choices(questions[qi], over_ans - 1, target - 1)
                        modified = True
                        fixed_one = True
                        break

            # Strategy 2: Find a swappable question with underrepresented ans,
            # change it to have ans=over_ans, freeing up a slot
            # (This doesn't help - it increases over_ans count)
            # Actually we need: find swappable with different ans, that is near a marker with over_ans,
            # swap positions, then change the swappable's ans
            if not fixed_one:
                # Find a marker question with over_ans and a swappable with different ans
                under_ans = min(dist, key=dist.get)
                for qi in range(len(questions)):
                    if (questions[qi].get('ans') == over_ans and
                        is_marker(questions[qi])):
                        # Find a swappable nearby with under_ans that we can change to over_ans
                        # Actually, we need to SWAP a marker(over_ans) with a swappable(different_ans)
                        # then change the swappable(now at marker's old position) to have a different ans
                        # Wait, that doesn't change anything.
                        #
                        # Better: find a swappable with any non-over ans, and change its choices
                        # so its ans becomes something else, reducing... no.
                        #
                        # The real fix: find ANY swappable question and change it FROM over_ans
                        # But if no swappable has over_ans, we're stuck.
                        pass

                # If truly stuck (all over_ans questions are markers), try position swaps
                # to at least group them differently
                # Actually, for A6 we just need fewer of over_ans.
                # If we can find a swappable question that currently has a different ans,
                # we could change it... but that increases its target's count.
                # The real fix needs to reduce over_ans count, which means changing a question
                # that has ans=over_ans. If all such questions are markers, we're stuck
                # unless we can change the marker's ans (which means changing which marker is wrong
                # in the passage - not possible).
                pass

        if not fixed_one:
            break

    if modified:
        data['questions'] = questions
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
            f.write('\n')

    return modified

def validate_file(filepath):
    result = subprocess.run(
        ['node', 'validate/validate.js', filepath],
        capture_output=True, text=True,
        cwd='/Users/woobumpark/Desktop/영어해방공식&내신핏/naesinfit-tests'
    )
    output = result.stdout + result.stderr
    first_line = output.strip().split('\n')[0] if output.strip() else ''
    has_a6a7 = bool(re.search(r'\[S\] A[67]:', output))
    return first_line, has_a6a7, output

if __name__ == '__main__':
    os.chdir('/Users/woobumpark/Desktop/영어해방공식&내신핏/naesinfit-tests')

    if len(sys.argv) > 1:
        files = sys.argv[1:]
    else:
        print('Finding FAIL files...')
        result = subprocess.run(
            ['bash', '-c', r'''find data/교과서 -maxdepth 10 \( -name "단어.json" -o -name "워크북.json" -o -name "퀴즈.json" \) | while IFS= read -r f; do result=$(node validate/validate.js "$f" 2>&1 | head -1); if echo "$result" | grep -q FAIL; then echo "$f"; fi; done'''],
            capture_output=True, text=True, timeout=600
        )
        files = [f for f in result.stdout.strip().split('\n') if f]

    total = len(files)
    fixed = 0
    still_fail = []

    for i, filepath in enumerate(files):
        print(f'[{i+1}/{total}] {filepath}')

        fl, has_a6a7, out = validate_file(filepath)
        if not has_a6a7:
            print(f'  -> No A6/A7, skip')
            continue

        was_modified = fix_file(filepath)
        if was_modified:
            fl2, has_a6a7_2, out2 = validate_file(filepath)
            if has_a6a7_2:
                print(f'  -> STILL A6/A7!')
                for line in out2.split('\n'):
                    if '[S] A6' in line or '[S] A7' in line:
                        print(f'     {line.strip()}')
                still_fail.append(filepath)
            else:
                print(f'  -> {fl2}')
                fixed += 1
        else:
            print(f'  -> No fix possible')
            still_fail.append(filepath)

    print(f'\n=== Summary ===')
    print(f'Total: {total}, Fixed: {fixed}, Still failing: {len(still_fail)}')
    if still_fail:
        for f in still_fail:
            print(f'  {f}')

File: fix-a6a7.js
#!/usr/bin/env node
/**
 * fix-a6a7.js — A6/A7 자동 수정 (det.analysis 동시 업데이트)
 *
 * mc 문항의 선지(ch)를 스왑하여 정답번호를 재분배합니다.
 * 마커형 문항(①②③④)은 고정, 나머지만 스왑.
 * det.analysis의 ①②③④ 라인과 ←정답 마커도 함께 업데이트.
 *
 * Usage:
 *   node fix-a6a7.js                          # data/교과서 전체
 *   node fix-a6a7.js <file1> <file2> ...      # 개별 파일
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const circleMarkers = ['①', '②', '③', '④', '⑤'];

function isMarkerType(q) {
  return q.fmt === 'mc' && Array.isArray(q.ch) && q.ch.length === 4 &&
    q.ch.every(c => circleMarkers.includes((c || '').trim()));
}

function isMc(q) {
  return q.fmt === 'mc' && typeof q.ans === 'number' && Array.isArray(q.ch) && q.ch.length === 4;
}

function isShuffleable(q) {
  return isMc(q) && !isMarkerType(q);
}

function hasA6A7(questions) {
  const mcQs = questions.filter(q => isMc(q));
  const dist = {};
  mcQs.forEach(q => { dist[q.ans] = (dist[q.ans] || 0) + 1; });
  if (Object.values(dist).some(c => c > 5)) return true;
  for (let i = 0; i < mcQs.length - 2; i++) {
    if (mcQs[i].ans === mcQs[i + 1].ans && mcQs[i + 1].ans === mcQs[i + 2].ans) return true;
  }
  return false;
}

/**
 * Swap two choices in a question, updating ch, ans, and det.analysis
 */
function swapChoices(q, idxA, idxB) {
  if (idxA === idxB) return;
  const ch = q.ch;

  // Swap choices
  const tmp = ch[idxA];
  ch[idxA] = ch[idxB];
  ch[idxB] = tmp;

  // Update ans
  if (q.ans - 1 === idxA) q.ans = idxB + 1;
  else if (q.ans - 1 === idxB) q.ans = idxA + 1;

  // Update det.analysis
  if (q.det && q.det.analysis) {
    const markers = ['①', '②', '③', '④'];
    const mA = markers[idxA];
    const mB = markers[idxB];

    const lines = q.det.analysis.split('\n');
    let lineA = null, lineB = null;
    let liA = -1, liB = -1;

    for (let i = 0; i < lines.length; i++) {
      const s = lines[i].trim();
      if (s.startsWith(mA)) { lineA = lines[i]; liA = i; }
      else if (s.startsWith(mB)) { lineB = lines[i]; liB = i; }
    }

    if (liA >= 0 && liB >= 0) {
      // Extract content after marker
      const contentA = lineA.includes(mA) ? lineA.split(mA).slice(1).join(mA) : '';
      const contentB = lineB.includes(mB) ? lineB.split(mB).slice(1).join(mB) : '';

      const arrow = ' ←정답';
      const cleanA = contentA.replace(arrow, '');
      const cleanB = contentB.replace(arrow, '');
      const aHadArrow = contentA.includes(arrow);
      const bHadArrow = contentB.includes(arrow);

      // After swap: position A gets B's content, position B gets A's content
      // Arrow follows the correct answer
      lines[liA] = mA + (bHadArrow ? cleanB : cleanB + (aHadArrow ? '' : ''));
      lines[liB] = mB + (aHadArrow ? cleanA : cleanA + (bHadArrow ? '' : ''));

      // Now add arrow to whichever has the correct content
      if (aHadArrow) {
        // A was correct, moved to B
        lines[liB] = lines[liB].replace(/\s*←정답\s*$/, '') + arrow;
        lines[liA] = lines[liA].replace(/\s*←정답\s*$/, '');
      } else if (bHadArrow) {
        // B was correct, moved to A
        lines[liA] = lines[liA].replace(/\s*←정답\s*$/, '') + arrow;
        lines[liB] = lines[liB].replace(/\s*←정답\s*$/, '');
      }

      q.det.analysis = lines.join('\n');
    }
  }
}

/**
 * Set a shuffleable question's answer to targetAns by swapping choices
 */
function setAnsTo(q, targetAns) {
  if (!isShuffleable(q)) return;
  if (q.ans === targetAns) return;

  const currentIdx = q.ans - 1;
  const targetIdx = targetAns - 1;
  swapChoices(q, currentIdx, targetIdx);
}

function fixA6A7(questions) {
  const allQs = questions;
  const mcPositions = [];

  allQs.forEach((q, i) => {
    if (isMc(q)) {
      mcPositions.push({ idx: i, fixed: isMarkerType(q), ans: q.ans });
    }
  });

  const n = mcPositions.length;
  if (n === 0) return false;

  // Phase 1: Generate target pattern (deterministic)
  function tryPattern(seed) {
    const targetAns = new Array(n);
    const count = { 1: 0, 2: 0, 3: 0, 4: 0 };

    // Fix marker positions
    mcPositions.forEach((p, i) => {
      if (p.fixed) {
        targetAns[i] = p.ans;
        count[p.ans]++;
      }
    });

    // Check if markers already violate A6
    if (Object.values(count).some(c => c > 5)) {
      // Can't fix if markers alone exceed 5
      // Try to still find valid pattern for shuffleable ones
    }

    const shuffleIndices = [];
    mcPositions.forEach((p, i) => {
      if (!p.fixed) shuffleIndices.push(i);
    });

    // Build pool
    const pool = [];
    [1, 2, 3, 4].forEach(a => {
      const remaining = 5 - count[a];
      for (let j = 0; j < Math.max(0, remaining); j++) pool.push(a);
    });

    while (pool.length < shuffleIndices.length) {
      const minAns = [1, 2, 3, 4].reduce((a, b) => {
        const ca = count[a] + pool.filter(x => x === a).length;
        const cb = count[b] + pool.filter(x => x === b).length;
        return ca <= cb ? a : b;
      });
      pool.push(minAns);
    }

    // Seed-based shuffle
    let s = seed;
    for (let i = pool.length - 1; i > 0; i--) {
      s = (s * 1664525 + 1013904223) & 0x7fffffff;
      const j = s % (i + 1);
      [pool[i], pool[j]] = [pool[j], pool[i]];
    }

    // Truncate pool if needed
    pool.length = shuffleIndices.length;

    shuffleIndices.forEach((si, j) => {
      targetAns[si] = pool[j];
    });

    // Validate: no 3 consecutive
    for (let i = 0; i < n - 2; i++) {
      if (targetAns[i] === targetAns[i + 1] && targetAns[i + 1] === targetAns[i + 2]) return null;
    }

    // Validate: distribution
    const finalCount = { 1: 0, 2: 0, 3: 0, 4: 0 };
    targetAns.forEach(a => finalCount[a]++);
    if (Object.values(finalCount).some(c => c > 5)) return null;

    return targetAns;
  }

  for (let seed = 0; seed < 20000; seed++) {
    const pattern = tryPattern(seed);
    if (pattern) {
      let changed = false;
      mcPositions.forEach((p, i) => {
        if (!p.fixed && pattern[i] !== allQs[p.idx].ans) {
          setAnsTo(allQs[p.idx], pattern[i]);
          changed = true;
        }
      });
      if (changed && !hasA6A7(allQs)) return true;
      if (changed) {
        // Revert (reload needed, but since we modify in place, this approach is problematic)
        // Instead, let's try another seed
        // We need to re-read the original - skip revert, the changes accumulate
        // which is fine as long as final validation passes
        continue;
      }
    }
  }

  // Phase 2: Position swap for marker-heavy A7 runs
  // If A7 remains because of marker-only consecutive runs,
  // swap question positions to break the run
  const mcQs = allQs.filter(q => isMc(q));
  for (let attempt = 0; attempt < 100; attempt++) {
    // Find first A7 violation
    let runFound = false;
    for (let i = 0; i < mcQs.length - 2; i++) {
      if (mcQs[i].ans === mcQs[i + 1].ans && mcQs[i + 1].ans === mcQs[i + 2].ans) {
        // Find a question outside this run with different ans
        const runVal = mcQs[i].ans;
        const midQ = mcQs[i + 1];
        const midGlobalIdx = allQs.indexOf(midQ);

        // Find swap candidate
        for (let j = 0; j < allQs.length; j++) {
          if (j === midGlobalIdx) continue;
          const candidate = allQs[j];
          if (!isMc(candidate)) continue;
          if (candidate.ans === runVal) continue;

          // Test swap
          const tmpQ = allQs[midGlobalIdx];
          allQs[midGlobalIdx] = allQs[j];
          allQs[j] = tmpQ;

          if (!hasA6A7(allQs)) return true;

          // Check if at least A7 improved
          const mcAfter = allQs.filter(q => isMc(q));
          let a7Count = 0;
          for (let k = 0; k < mcAfter.length - 2; k++) {
            if (mcAfter[k].ans === mcAfter[k + 1].ans && mcAfter[k + 1].ans === mcAfter[k + 2].ans) a7Count++;
          }

          // If no improvement, revert
          if (a7Count >= 1) {
            allQs[j] = allQs[midGlobalIdx];
            allQs[midGlobalIdx] = tmpQ;
          } else {
            runFound = true;
            break;
          }
        }
        if (runFound) break;
      }
    }
    if (!runFound) break;
  }

  return !hasA6A7(allQs);
}

function processFile(filepath) {
  let data;
  try {
    data = JSON.parse(fs.readFileSync(filepath, 'utf8'));
  } catch (e) {
    return { status: 'skip', reason: 'parse error' };
  }

  if (!Array.isArray(data.questions) || data.questions.length === 0) {
    return { status: 'skip', reason: 'no questions' };
  }

  if (!hasA6A7(data.questions)) {
    return { status: 'skip', reason: 'no A6/A7' };
  }

  // Deep copy for safety
  const backup = JSON.stringify(data);

  const fixed = fixA6A7(data.questions);

  if (!hasA6A7(data.questions)) {
    fs.writeFileSync(filepath, JSON.stringify(data, null, 2) + '\n', 'utf8');
    return { status: 'fixed' };
  } else {
    // Restore backup
    fs.writeFileSync(filepath, backup, 'utf8');
    // Get remaining errors
    const mcQs = data.questions.filter(q => isMc(q));
    const dist = {};
    mcQs.forEach(q => { dist[q.ans] = (dist[q.ans] || 0) + 1; });
    const fixedCount = mcQs.filter(q => isMarkerType(q)).length;
    return { status: 'fail', detail: `mc:${mcQs.length} markers:${fixedCount} dist:${JSON.stringify(dist)}` };
  }
}

function validate(filepath) {
  try {
    const out = execSync(`node validate/validate.js "${filepath}" 2>&1`, {
      cwd: '/Users/woobumpark/Desktop/영어해방공식&내신핏/naesinfit-tests',
      encoding: 'utf8',
      timeout: 30000
    });
    return out.trim().split('\n')[0];
  } catch (e) {
    return (e.stdout || e.stderr || '').trim().split('\n')[0];
  }
}

// ── Main ──
const args = process.argv.slice(2);
let files = [];

if (args.length > 0) {
  files = args;
} else {
  // Find FAIL files
  console.log('Finding FAIL files in data/교과서 ...');
  try {
    const out = execSync(
      `find data/교과서 -maxdepth 10 \\( -name "단어.json" -o -name "워크북.json" -o -name "퀴즈.json" \\) -print0 | xargs -0 -I{} sh -c 'r=$(node validate/validate.js "{}" 2>&1 | head -1); echo "$r" | grep -q FAIL && echo "{}"'`,
      { encoding: 'utf8', timeout: 600000, maxBuffer: 10 * 1024 * 1024,
        cwd: '/Users/woobumpark/Desktop/영어해방공식&내신핏/naesinfit-tests' }
    );
    files = out.trim().split('\n').filter(Boolean);
  } catch (e) {
    const out = (e.stdout || '').trim();
    files = out.split('\n').filter(Boolean);
  }
}

let fixed = 0, failed = 0, skipped = 0;
const failList = [];

files.forEach((f, i) => {
  process.stdout.write(`[${i + 1}/${files.length}] ${f} ... `);

  const result = processFile(f);

  if (result.status === 'fixed') {
    // Validate with validate.js
    const vResult = validate(f);
    const hasA6A7Still = /\[S\] A[67]:/.test(vResult);
    if (hasA6A7Still) {
      console.log(`STILL FAIL: ${vResult}`);
      failed++;
      failList.push(f);
    } else {
      console.log(`FIXED → ${vResult}`);
      fixed++;
    }
  } else if (result.status === 'fail') {
    console.log(`FAIL (${result.detail})`);
    failed++;
    failList.push(f);
  } else {
    console.log(`skip (${result.reason})`);
    skipped++;
  }
});

console.log(`\n=== Summary ===`);
console.log(`Total: ${files.length}, Fixed: ${fixed}, Failed: ${failed}, Skipped: ${skipped}`);
if (failList.length > 0) {
  console.log('Still failing:');
  failList.forEach(f => console.log(`  ${f}`));
}
